sqlglot/tokens.py

```diff
@@ -1361,7 +1365,9 @@ def _scan_string(self, start: str) -> bool:
 
     def _scan_identifier(self, identifier_end: str) -> None:
         self._advance()
-        text = self._extract_string(identifier_end, escapes=self._IDENTIFIER_ESCAPES)
+        text = self._extract_string(
+            identifier_end, escapes=self._IDENTIFIER_ESCAPES | {identifier_end}
```

Comment:
this seems expensive, we shouldn't be doing a set comparison inside this hot loop

Reply:
One alternative is to push the complexity down to `_extract_string`, which would mean:

- Adding a new flag so that we know when to treat `identifier_end` (i.e. its `delimiter` argument) as an escape
- Updating the [two places](https://github.com/tobymao/sqlglot/blob/main/sqlglot/tokens.py#L1406-L1407) where we check `escapes` to include an `or` operation to check if the `_char` / `_peek` matches `delimiter` as a special case.

The performance tradeoff is not obvious to me in this case. Should I go ahead and benchmark the two implementations? Something like tokenizing a string with many identifiers in it.

Reply:
ya, can you bench mark it

Reply:
Didn't notice any substantial differences between the two implementations. This is the benchmark script:

```
(.venv) ➜  sqlglot git:(jo/fix_identifier_delimiter_escaping) ✗ cat tokenize_identifiers.py
from timeit import timeit
from sqlglot import tokenize

sql = "SELECT " + ", ".join(["x"] * 1000)

print(timeit(lambda: tokenize(sql), number=1000))
```

Using the code in this PR:

```
(.venv) ➜  sqlglot git:(jo/fix_identifier_delimiter_escaping) ✗ SQLGLOTRS_TOKENIZER=0 python tokenize_identifiers.py
3.1507482500746846
(.venv) ➜  sqlglot git:(jo/fix_identifier_delimiter_escaping) ✗ SQLGLOTRS_TOKENIZER=1 python tokenize_identifiers.py
8.340694999787956
```

... and using the alternative implementation:

```
(.venv) ➜  sqlglot git:(jo/fix_identifier_delimiter_escaping) ✗ SQLGLOTRS_TOKENIZER=0 python tokenize_identifiers.py
3.189114083070308
(.venv) ➜  sqlglot git:(jo/fix_identifier_delimiter_escaping) ✗ SQLGLOTRS_TOKENIZER=1 python tokenize_identifiers.py
8.401180792134255
(.venv) ➜  sqlglot git:(jo/fix_identifier_delimiter_escaping) ✗
```

<details>
<summary>Click this to see the code diff</summary>
```diff
ns.py
+++ b/sqlglot/tokens.py
@@ -1366,7 +1366,7 @@ class Tokenizer(metaclass=_Tokenizer):
     def _scan_identifier(self, identifier_end: str) -> None:
         self._advance()
         text = self._extract_string(
-            identifier_end, escapes=self._IDENTIFIER_ESCAPES | {identifier_end}
+            identifier_end, escapes=self._IDENTIFIER_ESCAPES, delimiter_is_escape=True
         )
         self._add(TokenType.IDENTIFIER, text)

@@ -1390,6 +1390,7 @@ class Tokenizer(metaclass=_Tokenizer):
         escapes: t.Optional[t.Set[str]] = None,
         raw_string: bool = False,
         raise_unmatched: bool = True,
+        delimiter_is_escape: bool = False,
     ) -> str:
         text = ""
         delim_size = len(delimiter)
@@ -1409,7 +1410,7 @@ class Tokenizer(metaclass=_Tokenizer):
                     continue
             if (
                 (self.STRING_ESCAPES_ALLOWED_IN_RAW_STRINGS or not raw_string)
-                and self._char in escapes
+                and ((delimiter_is_escape and self._char == delimiter) or self._char in escapes)
                 and (self._peek == delimiter or self._peek in escapes)
                 and (self._char not in self._QUOTES or self._char == self._peek)
             ):
diff --git a/sqlglotrs/src/tokenizer.rs b/sqlglotrs/src/tokenizer.rs
index beff96e2..14388026 100644
--- a/sqlglotrs/src/tokenizer.rs
+++ b/sqlglotrs/src/tokenizer.rs
@@ -424,7 +424,7 @@ impl<'a> TokenizerState<'a> {
                 let tag = if self.current_char.to_string() == *end {
                     String::from("")
                 } else {
-                    self.extract_string(end, false, true, !self.settings.heredoc_tag_is_identifier)?
+                    self.extract_string(end, false, true, !self.settings.heredoc_tag_is_identifier, false)?
                 };

                 if !tag.is_empty()
@@ -449,7 +449,7 @@ impl<'a> TokenizerState<'a> {
         };

         self.advance(start.len() as isize)?;
-        let text = self.extract_string(&end, false, token_type == self.token_types.raw_string, true)?;
+        let text = self.extract_string(&end, false, token_type == self.token_types.raw_string, true, false)?;

         if let Some(b) = base {
             if u64::from_str_radix(&text, b).is_err() {
@@ -595,7 +595,7 @@ impl<'a> TokenizerState<'a> {

     fn scan_identifier(&mut self, identifier_end: &str) -> Result<(), TokenizerError> {
         self.advance(1)?;
-        let text = self.extract_string(identifier_end, true, false, true)?;
+        let text = self.extract_string(identifier_end, true, false, true, true)?;
         self.add(self.token_types.identifier, Some(text))
     }

@@ -605,19 +605,19 @@ impl<'a> TokenizerState<'a> {
         use_identifier_escapes: bool,
         raw_string: bool,
         raise_unmatched: bool,
+        delimiter_is_escape: bool,
     ) -> Result<String, TokenizerError> {
         let mut text = String::from("");

         loop {
-            let mut new_identifier_escapes;
+            // let mut new_identifier_escapes;
             let escapes = if use_identifier_escapes {
-                new_identifier_escapes = self.settings.identifier_escapes.clone();
-                new_identifier_escapes.extend(delimiter.chars());
-                &new_identifier_escapes
+                &self.settings.identifier_escapes
             } else {
                 &self.settings.string_escapes
             };
             let peek_char_str = self.peek_char.to_string();
+            let current_char_str = self.current_char.to_string();

             if !raw_string
                 && !self.dialect_settings.unescaped_sequences.is_empty()
@@ -635,13 +635,13 @@ impl<'a> TokenizerState<'a> {
             }

             if (self.settings.string_escapes_allowed_in_raw_strings || !raw_string)
-                && escapes.contains(&self.current_char)
+                && ((delimiter_is_escape && current_char_str == delimiter) || escapes.contains(&self.current_char))
                 && (peek_char_str == delimiter || escapes.contains(&self.peek_char))
                 && (self.current_char == self.peek_char
                     || !self
                         .settings
                         .quotes
-                        .contains_key(&self.current_char.to_string()))
+                        .contains_key(&current_char_str))
             {
                 if peek_char_str == delimiter {
                     text.push(self.peek_char);
```
</details>

As you can see, doing it the other way is even slightly slower in this particular example.